hypothesis-python/tests/watchdog/test_database.py

```diff
@@ -0,0 +1,152 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Copyright the Hypothesis Authors.
+# Individual contributors are listed in AUTHORS.rst and the git log.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+import sys
+import time
+
+from hypothesis import Phase, settings
+from hypothesis.database import (
+    DirectoryBasedExampleDatabase,
+    InMemoryExampleDatabase,
+    MultiplexedDatabase,
+)
+
+from tests.cover.test_database_backend import _database_conforms_to_listener_api
+
+# we need real time here, not monkeypatched for CI
+time_sleep = time.sleep
+
+
+def test_database_listener_directory():
+    # this test is very expensive because we wait between every rule for the
+    # filesystem observer to fire. Limit examples/step count as much as possible.
+    _database_conforms_to_listener_api(
+        lambda path: DirectoryBasedExampleDatabase(path),
+        flush=lambda _db: time_sleep(0.2),
+        supports_value_delete=False,
+        # expensive flush makes shrinking take forever
+        parent_settings=settings(
+            max_examples=5, stateful_step_count=10, phases=set(Phase) - {Phase.shrink}
+        ),
+    )
+
+
+def test_database_listener_multiplexed(tmp_path):
+    db = MultiplexedDatabase(
+        InMemoryExampleDatabase(), DirectoryBasedExampleDatabase(tmp_path)
+    )
+    events = []
+
+    def listener(event):
+        events.append(event)
+
+    db.add_listener(listener)
+
+    db.save(b"a", b"a")
+    time_sleep(0.2)
+    assert events == [("save", (b"a", b"a"))] * 2
+
+    db.remove_listener(listener)
+    db.delete(b"a", b"a")
+    db.save(b"a", b"b")
+    time_sleep(0.2)
+    assert events == [("save", (b"a", b"a"))] * 2
+
+    db.add_listener(listener)
+    db.delete(b"a", b"b")
+    db.save(b"a", b"c")
+    time_sleep(0.2)
+    # InMemory database fires immediately, while DirectoryBased has to
+    # wait for filesystem listeners. Therefore the events can arrive out of
+    # order. Test a weaker set property, disregarding ordering.
+    assert set(events[2:]) == {
```

Comment:
Hmm, since the `save` events are identical we might want a `Counter` here to get the multiset property?